### ISSUE TITLE
Cannot print/summarise model that was fit on cluster on Mac

### ISSUE THREAD

**Issue by rubenarslan** [stan-dev/rstan#543]
#### Summary:
I am fitting models via brms on a Torque cluster. Models are compiled on the cluster. When I download them to my Mac, I recently cannot call `summary` or `print` on them without crashing R.

#### Description:
I fit models on a cluster. Compilation happens there. There is a slight R version mismatch, but the package versions are the same. I didn't use to have this problem, but am not sure whether it happened after an rstan or R update.
[This problem](https://twitter.com/fusaroli/status/991220974048641025) sounds similar, but without a cluster being involved.

#### Reproducible Steps:
In the summary, I describe how I noticed the problem. But the following steps where sufficient to reproduce:

1. Run rstan example on Linux cluster with 3.3.3
```r
library(rstan)
scode <- "
parameters {
real y[2]; 
} 
model {
y[1] ~ normal(0, 1);
y[2] ~ double_exponential(0, 2);
} 
"
fit <- stan(model_code = scode, iter = 10, verbose = FALSE) 
saveRDS(fit, "fit.rds")
```

Transfer file to my Mac, then do `readRDS("fit.rds")`.

This leads to R crashing. With this example, I just get the crash and can see it's MALLOC in the Apple Crash report. When facing the same problem with brms, I can see the below error message.

I can summarise fit fine on the cluster, or when fitting the same model on the Mac. I haven't tried transferring a fit from Mac to the cluster.

#### Current Output:
> rsession(26725,0x7fffd7ac33c0) malloc: *** error for object 0x13da81968: incorrect checksum for freed object - object was probably modified after being freed.
*** set a breakpoint in malloc_error_break to debug

Apple Message

> Crashed Thread:        0  Dispatch queue: com.apple.main-thread
>
> Exception Type:        EXC_BAD_ACCESS (SIGABRT)
Exception Codes:       KERN_INVALID_ADDRESS at 0x0000610000600000
Exception Note:        EXC_CORPSE_NOTIFY
>
> Termination Reason:    Namespace OBJC, Code 0x1
>
> VM Regions Near 0x610000600000:
    MALLOC_NANO            0000610000000000-0000610000600000 [ 6144K] rw-/rwx SM=PRV  
--> 
    MALLOC_NANO            0000618000000000-0000618000800000 [ 8192K] rw-/rwx SM=PRV  
>
> Application Specific Information:
objc[33958]: Method cache corrupted. This may be a message to an invalid object, or a memory error somewhere else.
objc[33958]: receiver 0x1029cab30, SEL 0x102992fac, isa 0x1029cab58, cache 0x1029cab68, buckets 0x61000009a4f0, mask 0x3, occupied 0x2
objc[33958]: receiver 0 bytes, buckets 80 bytes
objc[33958]: selector 'mainHandler'
objc[33958]: isa 'REngine'

#### Expected Output:
model summary. I can call e.g. `fixef` on it though without crashing.

#### RStan Version:
> packageVersion('rstan')
[1] ‘2.17.3’
> packageVersion('brms')
[1] ‘2.3.1’

#### R Version:
The version of R you are running (e.g., from `R.version.string`)
> sessionInfo()
R version 3.4.4 RC (2018-03-08 r74373)

#### Operating System:
Your operating system (e.g., OS X 10.11.3)
macOS Sierra 10.12.6

#### On cluster
Same package versions, but:
R version 3.3.3 (2017-03-06)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Debian GNU/Linux 9 (stretch)

**Comment by bob-carpenter** [comment]
Binaries won't be compatible across platforms.

**Comment by rubenarslan** [comment]
Sorry, I don't understand this explanation. Do you mean across R versions or across Linux/Mac? So it is not supported to fit a model on a cluster and e.g. summarise on it on a different machine? Because this worked fine in the past (and I can extract all the parts of the summary that I care about without causing a crash). 
So, I don't quite know what the model binary is even used for when calling summary/print. 
@paul-buerkner, can you translate?

**Comment by maverickg** [comment]
`summary` and `print` should work if the rds is compatible across platforms and different version of R, which seems to be.  It seems here the problem occurred during reading the rds file---`print` and `summary` have not been called yet.  If that's the case, make sure you rds files are good first.

**Comment by bob-carpenter** [comment]
> it is not supported to fit a model on a cluster and e.g. summarise on it on a different machine?

Correct.  Unless the machines are binary compatible with their C++ compiler.

>  I don't quite know what the model binary is even used for when calling summary/print.

Everything is packed into that binary including the compiled model.  It's the compiled model that's causing problems.  If you just extract the draws, that should be serializable and you can use them to rebuild a kind of fit object.  That will work cross-OS.

If you want more precise instructions, please ask on our Discourse forums---this is the issue tracker and not a discussion forum.

**Comment by rubenarslan** [comment]
@maverickg I just wanted to make the example minimal. I can read the RDS-file fine (and like I said, I can extract draws and many other slots of the object). I can't call summary on it.

@bob-carpenter Thanks, I don't need instructions to extract draws and did not ask that. You did not answer my question why the compiled model is touched for summary. 
It seems like a common use case that doesn't work, even though it used to. It also seems to be bad error handling to just let this crash the R session. Seems to me this is an issue (as I don't need instructions), but happy to move to Discourse if you prefer.

**Comment by bob-carpenter** [comment]
I'll butt out and re-open the issue and let the RStan devs try to sort it out.

**Comment by maverickg** [comment]
I tried this and it's still ok for me.  That is, I used the above code to create a fit.rds file on linux and loaded on Mac.  Besides, it's only you cannot use this object to do sampling/optimization; even in this case, it would report the issue. 

Output on Mac:
```
 a <- readRDS('fit.rds')
> a
Inference for Stan model: bda3440e57006223a5830043c2b817ec.
4 chains, each with iter=10; warmup=5; thin=1;
post-warmup draws per chain=5, total post-warmup draws=20.

      mean se_mean   sd  2.5%   25%   50%   75% 97.5% n_eff Rhat
y[1]  0.30    0.25 0.97 -0.99 -0.66  0.25  1.11  1.69    15 2.31
y[2] -0.79    0.40 0.97 -2.09 -1.35 -0.86 -0.56  1.21     6 1.25
lp__ -1.05    0.12 0.45 -1.90 -1.29 -0.91 -0.70 -0.41    14 2.03

Samples were drawn using NUTS(diag_e) at Thu Jun 28 22:01:17 2018.
For each parameter, n_eff is a crude measure of effective sample size,
and Rhat is the potential scale reduction factor on split chains (at
convergence, Rhat=1).
> stan(fit=a)
Error in prep_call_sampler(object) :
  the compiled object from C++ code for this model is invalid, possible reasons:
  - compiled with save_dso=FALSE;
  - compiled on a different platform;
  - does not exist (created from reading csv files).
>
>
> a@stanmodel@dso
S4 class cxxdso: dso_saved = TRUE, dso_filename = fileb13bb3c9ff, size = 23.7 Mb.
And dso_last_path = '/tmp/RtmpHi5FX5/fileb13bb3c9ff.so'.
Created on: 'x86_64, linux-gnu' with 'CXXFLAGS = -g -O2 -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security -D_FORTIFY_SOURCE=2 -g $(LTO)'.
Loaded now: NO.
The signatures is/are as follows:
$fileb13bb3c9ff
character(0)
```
So I don't really know why it does not work for you @rubenarslan .

**Comment by rubenarslan** [comment]
@maverickg how can I help you reproduce this/narrow it down?
Like I said, I used to have no problems with Linux to Mac as long as it was just about getting summaries/draws, etc (can't refit models fit on the cluster on the smaller Mac anyway). So, I suspect it's something specific about the combination of these R versions and am not surprised you don't get the same error (if everyone experienced this, you'd have more complaints).

I can report the contents of `a@stanmodel@dso`

> S4 class cxxdso: dso_saved = TRUE, dso_filename = filebb2c4d81b03, size = 26.3 Mb.
And dso_last_path = '/tmp/Rtmpyy3CqG/filebb2c4d81b03.so'.
Created on: 'x86_64, linux-gnu' with 'CXXFLAGS = -g -O2 -fdebug-prefix-map=/build/r-base-3.3.3=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g $(LTO)'.
Loaded now: NO.
The signatures is/are as follows: 
$filebb2c4d81b03
character(0)

I can report that I also get a "nice" error when calling `stan(fit = a)`. A crash only happens when printing or more specifically when calling the summary method for stanfits. I had to look up how to debug S4 methods. This is what I did:

```r
library(rstan)
a <- readRDS("who_fits/fit.rds")
debug(summary, signature = "stanfit")
summary(a)
```

This way I found out that the crash happens when calling `rstan:::summary_sim(a@sim)`.

So

```r
debug(rstan:::summary_sim)
rstan:::summary_sim(a@sim)
```

in there it's `rstan_ess` that causes the crash. I don't know how to debug Cpp, so I can't make it more minimal. `rstan:::rstan_ess(a@sim, 1)` reproduces the crash for me. I'll upload the fit here in the hope that that helps.
[fit.rds.zip](https://github.com/stan-dev/rstan/files/2148181/fit.rds.zip)

**Comment by maverickg** [comment]
which version of rstan are you using? @rubenarslan ?

there was a bug that could cause this.  But I think it's only introduced on develop branch. https://github.com/stan-dev/rstan/issues/540

**Comment by rubenarslan** [comment]
I have 2.17.3.  I was fairly sure that I installed this from CRAN, because I updated all packages from CRAN at least once to see if that fixed the problem. Probably, I accidentally installed rstan from Github when I installed brms from Github, and update.packages didn't catch it because it was the same version number.
Either way, reinstalling from CRAN solved the problem. Thanks for your patience. I'm sure I should have noticed that earlier, but it would have been easier to notice, if you bumped the version number for the dev version. Hadley recommends adding [.9000](http://r-pkgs.had.co.nz/release.html). Maybe that would save you time in the long run :smile: 

Thanks for the quick support.
